Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationResultDispatchingValidator.cs

```diff
@@ -36,16 +36,79 @@ public void DispatchValidationFailures (IBusinessObjectValidationResult validati
       ArgumentUtility.CheckNotNull("validationResult", validationResult);
 
       var bocListControl = GetControlToValidate();
+      var useNewValidation = bocListControl
+          .GetColumnDefinitions()
+          .Any(cd => cd is BocValidationErrorIndicatorColumnDefinition);
 
+      if (useNewValidation)
+        DispatchNewImplementation(validationResult, bocListControl);
+      else
+        DispatchPreviousImplementation(validationResult, bocListControl);
+    }
+
+    //TODO Really need some better name here, but I don't know what the exact difference is so...help please.
+    private void DispatchPreviousImplementation (IBusinessObjectValidationResult validationResult, BocList bocListControl)
+    {
       var validatorsMatchingToControls = EnumerableUtility.SelectRecursiveDepthFirst(
               bocListControl as Control,
-              child => child.Controls.Cast<Control>().Where(item => !(item is INamingContainer)))
+              child => child.Controls.Cast<Control>().Where(item => item is not INamingContainer))
           .OfType<IBusinessObjectBoundEditableWebControlValidationResultDispatcher>();
 
       foreach (var validator in validatorsMatchingToControls)
         validator.DispatchValidationFailures(validationResult);
     }
 
+    //TODO Really need some better name here, but I don't know what the exact difference is so...help please.
+    private void DispatchNewImplementation (IBusinessObjectValidationResult validationResult, BocList bocListControl)
+    {
+      var rowObjects = bocListControl.Value?.Cast<IBusinessObject>();
+      if (rowObjects == null)
+        return;
+
+      var columnsAndMatchers = bocListControl
+          .GetColumnDefinitions()
+          .OfType<IBocColumnDefinitionWithValidationSupport>()
+          .Select(c => new { Column = c, Matcher = c.GetValidationFailureMatcher() })
+          .ToArray();
+
+      var validationFailureRepository = bocListControl.ValidationFailureRepository;
+      validationFailureRepository.ClearAllValidationFailures();
+
+      foreach (var rowObject in rowObjects)
+      {
+        foreach (var columnAndMatcher in columnsAndMatchers)
+        {
+          var matched = columnAndMatcher.Matcher.GetMatchingValidationFailures(rowObject, validationResult);
+          BocColumnDefinition columnDefinition = columnAndMatcher.Column switch
```

Comment:
We don't need to do any fancy checks here we can cast to `BocColumnDefinition` directly. Even better, we can add the cast when we create the anonymous object `new { Column = (BocColumnDefinition)c, ... }`. We know that this cast is guaranteed to succeed as `.GetColumnDefinitions()` returns an `IReadOnlyList<BocColumnDefinition>`.